pkg/streamingpromql/operators/aggregations/topkbottomk/instant_query.go

```diff
@@ -0,0 +1,320 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package topkbottomk
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/aggregations"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// InstantQuery implements topk() and bottomk() for range queries.
+type InstantQuery struct {
+	Inner                    types.InstantVectorOperator
+	Param                    types.ScalarOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, New will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	IsTopK                   bool // If false, this is operator is for bottomk().
+
+	expressionPosition posrange.PositionRange
+	limit              int64 // Maximum number of values to return for each group.
+
+	values          []float64
+	nextSeriesIndex int
+
+	annotations                            *annotations.Annotations
+	haveEmittedIgnoredHistogramsAnnotation bool
+
+	// Reuse the same heap instance to allow us to avoid allocating a new one every time.
+	heap *instantQueryHeap
+}
+
+var _ types.InstantVectorOperator = &InstantQuery{}
+
+func (t *InstantQuery) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	if err := t.loadLimit(ctx); err != nil {
+		return nil, err
+	}
+
+	if t.limit == 0 {
+		// We can't return any series, so stop now.
+		return nil, nil
+	}
+
+	innerSeries, err := t.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(innerSeries)
+
+	groupLabelsBytesFunc := t.groupLabelsBytesFunc()
+	groups := map[string]*instantQueryGroup{}
+	seriesToGroups := make([]*instantQueryGroup, 0, len(innerSeries))
+
+	for _, series := range innerSeries {
+		groupLabelsString := groupLabelsBytesFunc(series.Labels)
+		g, groupExists := groups[string(groupLabelsString)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !groupExists {
+			g = &instantQueryGroup{}
+			groups[string(groupLabelsString)] = g
+		}
+
+		g.seriesCount++
+		seriesToGroups = append(seriesToGroups, g)
+	}
+
+	outputSeriesCount := 0
+
+	for idx, series := range innerSeries {
```

Comment:
Could this not be part of the loop just above? And do we even need `seriesToGroups`?

Reply:
> Could this not be part of the loop just above?

No, we need two loops here so that we have `seriesCount` for each group - this allows us to allocate each group's `series` slice appropriately. I'll add a comment to make this clearer.

> And do we even need `seriesToGroups`?

Yes, this allows us to avoid recomputing the group for each series in the second loop.

Reply:
Comment added in fdeb024

Reply:
Ah, I see. Thanks.